File: src/features/technical_indicators.py
"""
Technical indicators for cryptocurrency analysis.
Includes momentum, trend, volatility, and volume indicators.
"""

import pandas as pd
import numpy as np
from typing import Dict, List, Optional, Tuple
import ta
from ta.trend import SMAIndicator, EMAIndicator, MACD, ADXIndicator
from ta.momentum import RSIIndicator, StochasticOscillator, WilliamsRIndicator
from ta.volatility import BollingerBands, AverageTrueRange
from ta.volume import VolumeWeightedAveragePrice, OnBalanceVolumeIndicator
import logging

logger = logging.getLogger(__name__)


class TechnicalIndicators:
    """
    Comprehensive technical analysis indicators for cryptocurrency data.
    """
    
    def __init__(self, data: pd.DataFrame):
        """
        Initialize with OHLCV data.
        
        Args:
            data: DataFrame with OHLCV columns (OPEN, HIGH, LOW, CLOSE, VOLUME)
        """
        self.data = data.copy()
        self.indicators = {}
        
    def add_all_indicators(self) -> pd.DataFrame:
        """
        Add all technical indicators to the data.
        
        Returns:
            DataFrame with all indicators added
        """
        logger.info("Adding all technical indicators...")
        
        # Trend indicators
        self._add_moving_averages()
        self._add_macd()
        
        # Momentum indicators
        self._add_rsi()
        self._add_stochastic()
        self._add_williams_r()
        
        # Volatility indicators
        self._add_bollinger_bands()
        self._add_atr()
        
        # Volume indicators
        self._add_volume_indicators()
        
        # Custom indicators
        self._add_custom_indicators()
        
        logger.info(f"Added {len(self.indicators)} technical indicators")
        return self.data
    
    def _add_moving_averages(self):
        """Add various moving averages."""
        # Simple Moving Averages
        for period in [5, 10, 20, 50, 100, 200]:
            sma = SMAIndicator(close=self.data['CLOSE'], window=period)
            self.data[f'SMA_{period}'] = sma.sma_indicator()
            self.indicators[f'SMA_{period}'] = f'Simple Moving Average ({period})'
        
        # Exponential Moving Averages
        for period in [12, 26, 50, 100]:
            ema = EMAIndicator(close=self.data['CLOSE'], window=period)
            self.data[f'EMA_{period}'] = ema.ema_indicator()
            self.indicators[f'EMA_{period}'] = f'Exponential Moving Average ({period})'
        
        # Moving Average Crossovers
        self.data['SMA_CROSS_20_50'] = np.where(
            self.data['SMA_20'] > self.data['SMA_50'], 1, -1
        )
        self.data['EMA_CROSS_12_26'] = np.where(
            self.data['EMA_12'] > self.data['EMA_26'], 1, -1
        )
    
    def _add_macd(self):
        """Add MACD indicator."""
        macd = MACD(close=self.data['CLOSE'])
        self.data['MACD'] = macd.macd()
        self.data['MACD_SIGNAL'] = macd.macd_signal()
        self.data['MACD_HISTOGRAM'] = macd.macd_diff()
        
        # MACD crossovers
        self.data['MACD_CROSS'] = np.where(
            self.data['MACD'] > self.data['MACD_SIGNAL'], 1, -1
        )
        
        self.indicators.update({
            'MACD': 'MACD Line',
            'MACD_SIGNAL': 'MACD Signal Line',
            'MACD_HISTOGRAM': 'MACD Histogram',
            'MACD_CROSS': 'MACD Crossover Signal'
        })
    
    def _add_rsi(self):
        """Add RSI indicator."""
        for period in [14, 21]:
            rsi = RSIIndicator(close=self.data['CLOSE'], window=period)
            self.data[f'RSI_{period}'] = rsi.rsi()
            self.indicators[f'RSI_{period}'] = f'Relative Strength Index ({period})'
        
        # RSI signals
        rsi_14 = self.data['RSI_14']
        self.data['RSI_OVERSOLD'] = np.where(rsi_14 < 30, 1, 0)
        self.data['RSI_OVERBOUGHT'] = np.where(rsi_14 > 70, 1, 0)
        self.data['RSI_DIVERGENCE'] = self._calculate_rsi_divergence()
    
    def _add_stochastic(self):
        """Add Stochastic Oscillator."""
        stoch = StochasticOscillator(
            high=self.data['HIGH'],
            low=self.data['LOW'],
            close=self.data['CLOSE']
        )
        self.data['STOCH_K'] = stoch.stoch()
        self.data['STOCH_D'] = stoch.stoch_signal()
        
        self.indicators.update({
            'STOCH_K': 'Stochastic %K',
            'STOCH_D': 'Stochastic %D'
        })
    
    def _add_williams_r(self):
        """Add Williams %R indicator."""
        williams_r = WilliamsRIndicator(
            high=self.data['HIGH'],
            low=self.data['LOW'],
            close=self.data['CLOSE']
        )
        self.data['WILLIAMS_R'] = williams_r.williams_r()
        self.indicators['WILLIAMS_R'] = 'Williams %R'
    
    def _add_bollinger_bands(self):
        """Add Bollinger Bands."""
        bb = BollingerBands(close=self.data['CLOSE'])
        self.data['BB_UPPER'] = bb.bollinger_hband()
        self.data['BB_MIDDLE'] = bb.bollinger_mavg()
        self.data['BB_LOWER'] = bb.bollinger_lband()
        self.data['BB_WIDTH'] = bb.bollinger_wband()
        self.data['BB_POSITION'] = bb.bollinger_pband()
        
        self.indicators.update({
            'BB_UPPER': 'Bollinger Bands Upper',
            'BB_MIDDLE': 'Bollinger Bands Middle',
            'BB_LOWER': 'Bollinger Bands Lower',
            'BB_WIDTH': 'Bollinger Bands Width',
            'BB_POSITION': 'Bollinger Bands Position'
        })
    
    def _add_atr(self):
        """Add Average True Range."""
        atr = AverageTrueRange(
            high=self.data['HIGH'],
            low=self.data['LOW'],
            close=self.data['CLOSE']
        )
        self.data['ATR'] = atr.average_true_range()
        self.indicators['ATR'] = 'Average True Range'
    
    def _add_volume_indicators(self):
        """Add volume-based indicators."""
        # On-Balance Volume
        obv = OnBalanceVolumeIndicator(
            close=self.data['CLOSE'],
            volume=self.data['VOLUME']
        )
        self.data['OBV'] = obv.on_balance_volume()
        
        # Volume Weighted Average Price
        vwap = VolumeWeightedAveragePrice(
            high=self.data['HIGH'],
            low=self.data['LOW'],
            close=self.data['CLOSE'],
            volume=self.data['VOLUME']
        )
        self.data['VWAP'] = vwap.volume_weighted_average_price()
        
        # Volume moving averages
        self.data['VOLUME_SMA_20'] = self.data['VOLUME'].rolling(window=20).mean()
        self.data['VOLUME_RATIO'] = self.data['VOLUME'] / self.data['VOLUME_SMA_20']
        
        self.indicators.update({
            'OBV': 'On-Balance Volume',
            'VWAP': 'Volume Weighted Average Price',
            'VOLUME_RATIO': 'Volume Ratio'
        })
    
    def _add_custom_indicators(self):
        """Add custom technical indicators."""
        # Price action features
        self.data['PRICE_RANGE'] = self.data['HIGH'] - self.data['LOW']
        self.data['BODY_SIZE'] = abs(self.data['CLOSE'] - self.data['OPEN'])
        self.data['UPPER_SHADOW'] = self.data['HIGH'] - np.maximum(self.data['OPEN'], self.data['CLOSE'])
        self.data['LOWER_SHADOW'] = np.minimum(self.data['OPEN'], self.data['CLOSE']) - self.data['LOW']
        
        # Momentum features
        self.data['MOMENTUM_5'] = self.data['CLOSE'] / self.data['CLOSE'].shift(5) - 1
        self.data['MOMENTUM_10'] = self.data['CLOSE'] / self.data['CLOSE'].shift(10) - 1
        self.data['MOMENTUM_20'] = self.data['CLOSE'] / self.data['CLOSE'].shift(20) - 1
        
        # Volatility features
        self.data['VOLATILITY_5'] = self.data['RETURN'].rolling(window=5).std()
        self.data['VOLATILITY_10'] = self.data['RETURN'].rolling(window=10).std()
        self.data['VOLATILITY_20'] = self.data['RETURN'].rolling(window=20).std()
        
        # Support and Resistance levels (simplified)
        self.data['SUPPORT_20'] = self.data['LOW'].rolling(window=20).min()
        self.data['RESISTANCE_20'] = self.data['HIGH'].rolling(window=20).max()
        
        # Trend strength
        self.data['TREND_STRENGTH'] = abs(self.data['SMA_20'] - self.data['SMA_50']) / self.data['SMA_50']
        
        self.indicators.update({
            'MOMENTUM_5': '5-day Momentum',
            'MOMENTUM_10': '10-day Momentum',
            'MOMENTUM_20': '20-day Momentum',
            'VOLATILITY_5': '5-day Volatility',
            'VOLATILITY_10': '10-day Volatility',
            'VOLATILITY_20': '20-day Volatility',
            'TREND_STRENGTH': 'Trend Strength'
        })
    
    def _calculate_rsi_divergence(self) -> pd.Series:
        """
        Calculate RSI divergence (simplified version).
        
        Returns:
            Series with divergence signals
        """
        rsi = self.data['RSI_14']
        price = self.data['CLOSE']
        
        # Look for divergence over 20 periods
        divergence = pd.Series(0, index=self.data.index)
        
        for i in range(20, len(self.data)):
            # Check for bullish divergence (price lower lows, RSI higher lows)
            if (price.iloc[i] < price.iloc[i-10] and 
                rsi.iloc[i] > rsi.iloc[i-10] and 
                rsi.iloc[i] < 40):
                divergence.iloc[i] = 1
            
            # Check for bearish divergence (price higher highs, RSI lower highs)
            elif (price.iloc[i] > price.iloc[i-10] and 
                  rsi.iloc[i] < rsi.iloc[i-10] and 
                  rsi.iloc[i] > 60):
                divergence.iloc[i] = -1
        
        return divergence
    
    def get_indicator_list(self) -> Dict[str, str]:
        """
        Get list of all indicators with descriptions.
        
        Returns:
            Dictionary mapping indicator names to descriptions
        """
        return self.indicators
    
    def get_feature_matrix(self, 
                          exclude_ohlcv: bool = True,
                          exclude_nan: bool = True) -> pd.DataFrame:
        """
        Get feature matrix for machine learning.
        
        Args:
            exclude_ohlcv: Whether to exclude OHLCV columns
            exclude_nan: Whether to exclude columns with NaN values
            
        Returns:
            Feature matrix DataFrame
        """
        features = self.data.copy()
        
        if exclude_ohlcv:
            ohlcv_cols = ['OPEN', 'HIGH', 'LOW', 'CLOSE', 'VOLUME']
            features = features.drop(columns=ohlcv_cols, errors='ignore')
        
        if exclude_nan:
            # Remove columns with too many NaN values (>50%)
            nan_threshold = len(features) * 0.5
            features = features.dropna(axis=1, thresh=nan_threshold)
            
            # Remove rows with any remaining NaN values
            features = features.dropna()
        
        return features
    
    def get_signal_indicators(self) -> pd.DataFrame:
        """
        Get indicators that generate trading signals.
        
        Returns:
            DataFrame with signal indicators
        """
        signal_cols = [
            'MACD_CROSS', 'RSI_OVERSOLD', 'RSI_OVERBOUGHT', 'RSI_DIVERGENCE',
            'SMA_CROSS_20_50', 'EMA_CROSS_12_26', 'STOCH_K', 'STOCH_D',
            'WILLIAMS_R', 'BB_POSITION', 'VOLUME_RATIO'
        ]
        
        available_cols = [col for col in signal_cols if col in self.data.columns]
        return self.data[available_cols]


def calculate_indicators(data: pd.DataFrame) -> pd.DataFrame:
    """
    Convenience function to calculate all technical indicators.
    
    Args:
        data: OHLCV DataFrame
        
    Returns:
        DataFrame with all indicators added
    """
    ti = TechnicalIndicators(data)
    return ti.add_all_indicators()


def get_momentum_features(data: pd.DataFrame) -> pd.DataFrame:
    """
    Extract momentum-related features.
    
    Args:
        data: DataFrame with technical indicators
        
    Returns:
        DataFrame with momentum features
    """
    momentum_cols = [
        'RSI_14', 'RSI_21', 'STOCH_K', 'STOCH_D', 'WILLIAMS_R',
        'MOMENTUM_5', 'MOMENTUM_10', 'MOMENTUM_20', 'MACD', 'MACD_SIGNAL'
    ]
    
    available_cols = [col for col in momentum_cols if col in data.columns]
    return data[available_cols]


def get_volatility_features(data: pd.DataFrame) -> pd.DataFrame:
    """
    Extract volatility-related features.
    
    Args:
        data: DataFrame with technical indicators
        
    Returns:
        DataFrame with volatility features
    """
    volatility_cols = [
        'ATR', 'BB_WIDTH', 'VOLATILITY_5', 'VOLATILITY_10', 'VOLATILITY_20',
        'PRICE_RANGE', 'BODY_SIZE'
    ]
    
    available_cols = [col for col in volatility_cols if col in data.columns]
    return data[available_cols]


def get_trend_features(data: pd.DataFrame) -> pd.DataFrame:
    """
    Extract trend-related features.
    
    Args:
        data: DataFrame with technical indicators
        
    Returns:
        DataFrame with trend features
    """
    trend_cols = [
        'SMA_20', 'SMA_50', 'SMA_200', 'EMA_12', 'EMA_26', 'EMA_50',
        'ADX', 'TREND_STRENGTH', 'SMA_CROSS_20_50', 'EMA_CROSS_12_26'
    ]
    
    available_cols = [col for col in trend_cols if col in data.columns]
    return data[available_cols]


if __name__ == "__main__":
    # Example usage
    from src.data.data_collector import CryptoDataCollector
    
    # Get sample data
    collector = CryptoDataCollector()
    btc_data = collector.get_ohlcv_data('BTC', period='6mo')
    
    # Calculate indicators
    ti = TechnicalIndicators(btc_data)
    data_with_indicators = ti.add_all_indicators()
    
    print(f"Original columns: {len(btc_data.columns)}")
    print(f"With indicators: {len(data_with_indicators.columns)}")
    print(f"Indicators added: {len(ti.get_indicator_list())}")
    
    # Get feature matrix
    features = ti.get_feature_matrix()
    print(f"Feature matrix shape: {features.shape}")
    
    # Get signal indicators
    signals = ti.get_signal_indicators()
    print(f"Signal indicators: {signals.columns.tolist()}") 

File: R/technical_indicators.R
#' Technical Indicators for Cryptocurrency Analysis
#' 
#' A comprehensive module for calculating technical indicators
#' using TTR and other R packages for financial analysis.
#' 
#' @author Crypto Analyst
#' @keywords technical analysis, indicators, cryptocurrency

library(TTR)
library(dplyr)
library(zoo)
library(xts)

#' TechnicalIndicators Class
#' 
#' Calculates various technical indicators for cryptocurrency analysis
#' 
#' @field data OHLCV data frame
#' @field indicators List of calculated indicators
#' 
#' @export
TechnicalIndicators <- setRefClass(
  "TechnicalIndicators",
  fields = list(
    data = "data.frame",
    indicators = "list"
  ),
  
  methods = list(
    
    #' Initialize the technical indicators calculator
    #' 
    #' @param data OHLCV data frame
    initialize = function(data) {
      data <<- data
      indicators <<- list()
    },
    
    #' Add all technical indicators
    #' 
    #' @return Data frame with all indicators
    add_all_indicators = function() {
      message("Calculating technical indicators...")
      
      # Trend indicators
      add_moving_averages()
      add_bollinger_bands()
      add_parabolic_sar()
      add_ichimoku()
      
      # Momentum indicators
      add_rsi()
      add_stochastic()
      add_williams_r()
      add_macd()
      add_cci()
      add_momentum()
      
      # Volume indicators
      add_volume_indicators()
      add_obv()
      add_vwap()
      
      # Volatility indicators
      add_atr()
      add_keltner_channels()
      
      # Support and resistance
      add_support_resistance()
      
      message(paste("Added", length(indicators), "technical indicators"))
      return(data)
    },
    
    #' Add moving averages
    add_moving_averages = function() {
      # Simple Moving Averages
      data$SMA_5 <<- SMA(data$Close, n = 5)
      data$SMA_10 <<- SMA(data$Close, n = 10)
      data$SMA_20 <<- SMA(data$Close, n = 20)
      data$SMA_50 <<- SMA(data$Close, n = 50)
      data$SMA_100 <<- SMA(data$Close, n = 100)
      data$SMA_200 <<- SMA(data$Close, n = 200)
      
      # Exponential Moving Averages
      data$EMA_12 <<- EMA(data$Close, n = 12)
      data$EMA_26 <<- EMA(data$Close, n = 26)
      data$EMA_50 <<- EMA(data$Close, n = 50)
      
      # Weighted Moving Average
      data$WMA_20 <<- WMA(data$Close, n = 20)
      
      indicators$moving_averages <<- c("SMA_5", "SMA_10", "SMA_20", "SMA_50", 
                                     "SMA_100", "SMA_200", "EMA_12", "EMA_26", 
                                     "EMA_50", "WMA_20")
    },
    
    #' Add Bollinger Bands
    add_bollinger_bands = function() {
      bb <- BBands(data$Close, n = 20, sd = 2)
      data$BB_Upper <<- bb[, "upper"]
      data$BB_Middle <<- bb[, "middle"]
      data$BB_Lower <<- bb[, "lower"]
      data$BB_Width <<- (data$BB_Upper - data$BB_Lower) / data$BB_Middle
      data$BB_Position <<- (data$Close - data$BB_Lower) / (data$BB_Upper - data$BB_Lower)
      
      indicators$bollinger_bands <<- c("BB_Upper", "BB_Middle", "BB_Lower", 
                                     "BB_Width", "BB_Position")
    },
    
    #' Add RSI (Relative Strength Index)
    add_rsi = function() {
      data$RSI_14 <<- RSI(data$Close, n = 14)
      data$RSI_21 <<- RSI(data$Close, n = 21)
      
      # RSI divergences
      data$RSI_Overbought <<- ifelse(data$RSI_14 > 70, 1, 0)
      data$RSI_Oversold <<- ifelse(data$RSI_14 < 30, 1, 0)
      
      indicators$rsi <<- c("RSI_14", "RSI_21", "RSI_Overbought", "RSI_Oversold")
    },
    
    #' Add Stochastic Oscillator
    add_stochastic = function() {
      stoch <- stoch(data$High, data$Low, data$Close, nFastK = 14, nFastD = 3, nSlowD = 3)
      data$Stoch_K <<- stoch[, "fastK"]
      data$Stoch_D <<- stoch[, "fastD"]
      data$Stoch_SlowD <<- stoch[, "slowD"]
      
      # Stochastic signals
      data$Stoch_Overbought <<- ifelse(data$Stoch_K > 80, 1, 0)
      data$Stoch_Oversold <<- ifelse(data$Stoch_K < 20, 1, 0)
      
      indicators$stochastic <<- c("Stoch_K", "Stoch_D", "Stoch_SlowD", 
                                "Stoch_Overbought", "Stoch_Oversold")
    },
    
    #' Add Williams %R
    add_williams_r = function() {
      data$Williams_R <<- WPR(data$High, data$Low, data$Close, n = 14)
      
      # Williams %R signals
      data$Williams_Overbought <<- ifelse(data$Williams_R > -20, 1, 0)
      data$Williams_Oversold <<- ifelse(data$Williams_R < -80, 1, 0)
      
      indicators$williams_r <<- c("Williams_R", "Williams_Overbought", "Williams_Oversold")
    },
    
    #' Add MACD
    add_macd = function() {
      macd_data <- MACD(data$Close, nFast = 12, nSlow = 26, nSig = 9)
      data$MACD_Line <<- macd_data[, "macd"]
      data$MACD_Signal <<- macd_data[, "signal"]
      data$MACD_Histogram <<- macd_data[, "histogram"]
      
      # MACD signals
      data$MACD_Bullish <<- ifelse(data$MACD_Line > data$MACD_Signal, 1, 0)
      data$MACD_Bearish <<- ifelse(data$MACD_Line < data$MACD_Signal, 1, 0)
      
      indicators$macd <<- c("MACD_Line", "MACD_Signal", "MACD_Histogram", 
                           "MACD_Bullish", "MACD_Bearish")
    },
    
    #' Add CCI (Commodity Channel Index)
    add_cci = function() {
      data$CCI <<- CCI(data$High, data$Low, data$Close, n = 20)
      
      # CCI signals
      data$CCI_Overbought <<- ifelse(data$CCI > 100, 1, 0)
      data$CCI_Oversold <<- ifelse(data$CCI < -100, 1, 0)
      
      indicators$cci <<- c("CCI", "CCI_Overbought", "CCI_Oversold")
    },
    
    #' Add Momentum
    add_momentum = function() {
      data$Momentum_10 <<- momentum(data$Close, n = 10)
      data$Momentum_20 <<- momentum(data$Close, n = 20)
      
      indicators$momentum <<- c("Momentum_10", "Momentum_20")
    },
    
    #' Add volume indicators
    add_volume_indicators = function() {
      # Volume moving averages
      data$Volume_SMA_20 <<- SMA(data$Volume, n = 20)
      data$Volume_SMA_50 <<- SMA(data$Volume, n = 50)
      
      # Volume ratio
      data$Volume_Ratio <<- data$Volume / data$Volume_SMA_20
      
      # Volume price trend
      data$VPT <<- VPT(data$Close, data$Volume)
      
      # Money flow index
      data$MFI <<- MFI(data$High, data$Low, data$Close, data$Volume, n = 14)
      
      indicators$volume <<- c("Volume_SMA_20", "Volume_SMA_50", "Volume_Ratio", 
                            "VPT", "MFI")
    },
    
    #' Add OBV (On Balance Volume)
    add_obv = function() {
      data$OBV <<- OBV(data$Close, data$Volume)
      
      # OBV moving average
      data$OBV_MA <<- SMA(data$OBV, n = 20)
      
      indicators$obv <<- c("OBV", "OBV_MA")
    },
    
    #' Add VWAP (Volume Weighted Average Price)
    add_vwap = function() {
      # Simple VWAP calculation
      data$VWAP <<- cumsum(data$Close * data$Volume) / cumsum(data$Volume)
      
      indicators$vwap <<- "VWAP"
    },
    
    #' Add ATR (Average True Range)
    add_atr = function() {
      atr_data <- ATR(data$High, data$Low, data$Close, n = 14)
      data$ATR <<- atr_data[, "atr"]
      data$ATR_Percent <<- data$ATR / data$Close * 100
      
      indicators$atr <<- c("ATR", "ATR_Percent")
    },
    
    #' Add Keltner Channels
    add_keltner_channels = function() {
      # Keltner Channels using ATR
      data$KC_Upper <<- data$EMA_20 + (2 * data$ATR)
      data$KC_Lower <<- data$EMA_20 - (2 * data$ATR)
      data$KC_Middle <<- data$EMA_20
      
      indicators$keltner_channels <<- c("KC_Upper", "KC_Middle", "KC_Lower")
    },
    
    #' Add Parabolic SAR
    add_parabolic_sar = function() {
      data$PSAR <<- SAR(data$High, data$Low, acceleration = 0.02, maximum = 0.2)
      
      # PSAR signals
      data$PSAR_Bullish <<- ifelse(data$Close > data$PSAR, 1, 0)
      data$PSAR_Bearish <<- ifelse(data$Close < data$PSAR, 1, 0)
      
      indicators$psar <<- c("PSAR", "PSAR_Bullish", "PSAR_Bearish")
    },
    
    #' Add Ichimoku Cloud
    add_ichimoku = function() {
      # Ichimoku components
      data$Ichimoku_Tenkan <<- (runMax(data$High, 9) + runMin(data$Low, 9)) / 2
      data$Ichimoku_Kijun <<- (runMax(data$High, 26) + runMin(data$Low, 26)) / 2
      data$Ichimoku_Senkou_A <<- (data$Ichimoku_Tenkan + data$Ichimoku_Kijun) / 2
      data$Ichimoku_Senkou_B <<- (runMax(data$High, 52) + runMin(data$Low, 52)) / 2
      data$Ichimoku_Chikou <<- lag(data$Close, 26)
      
      indicators$ichimoku <<- c("Ichimoku_Tenkan", "Ichimoku_Kijun", "Ichimoku_Senkou_A", 
                               "Ichimoku_Senkou_B", "Ichimoku_Chikou")
    },
    
    #' Add support and resistance levels
    add_support_resistance = function() {
      # Rolling support and resistance
      data$Support_20 <<- rollapply(data$Low, width = 20, FUN = min, fill = NA, align = "right")
      data$Resistance_20 <<- rollapply(data$High, width = 20, FUN = max, fill = NA, align = "right")
      
      # Pivot points
      data$Pivot_Point <<- (data$High + data$Low + data$Close) / 3
      data$R1 <<- 2 * data$Pivot_Point - data$Low
      data$S1 <<- 2 * data$Pivot_Point - data$High
      data$R2 <<- data$Pivot_Point + (data$High - data$Low)
      data$S2 <<- data$Pivot_Point - (data$High - data$Low)
      
      indicators$support_resistance <<- c("Support_20", "Resistance_20", "Pivot_Point", 
                                        "R1", "S1", "R2", "S2")
    },
    
    #' Get feature matrix for machine learning
    #' 
    #' @return Matrix of technical indicators
    get_feature_matrix = function() {
      # Get all indicator columns
      all_indicators <- unlist(indicators)
      available_indicators <- all_indicators[all_indicators %in% names(data)]
      
      # Create feature matrix
      feature_matrix <- data[, available_indicators, drop = FALSE]
      
      # Remove rows with NA values
      feature_matrix <- feature_matrix[complete.cases(feature_matrix), , drop = FALSE]
      
      return(feature_matrix)
    },
    
    #' Get list of all indicators
    #' 
    #' @return List of indicator names
    get_indicator_list = function() {
      return(unlist(indicators))
    },
    
    #' Get indicator summary statistics
    #' 
    #' @return Data frame with summary statistics
    get_indicator_summary = function() {
      all_indicators <- unlist(indicators)
      available_indicators <- all_indicators[all_indicators %in% names(data)]
      
      if (length(available_indicators) == 0) {
        return(data.frame())
      }
      
      summary_stats <- data.frame(
        Indicator = available_indicators,
        Mean = sapply(data[available_indicators], mean, na.rm = TRUE),
        SD = sapply(data[available_indicators], sd, na.rm = TRUE),
        Min = sapply(data[available_indicators], min, na.rm = TRUE),
        Max = sapply(data[available_indicators], max, na.rm = TRUE),
        NAs = sapply(data[available_indicators], function(x) sum(is.na(x)))
      )
      
      return(summary_stats)
    }
  )
)

#' Example usage of technical indicators
#' 
#' @examples
#' # Load data
#' collector <- CryptoDataCollector()
#' btc_data <- collector$get_ohlcv_data("BTC-USD", period = "6mo")
#' 
#' # Calculate indicators
#' ti <- TechnicalIndicators(btc_data)
#' data_with_indicators <- ti$add_all_indicators()
#' 
#' # Get feature matrix
#' features <- ti$get_feature_matrix()
#' 
#' # Get summary
#' summary <- ti$get_indicator_summary()
#' print(summary)
example_technical_indicators <- function() {
  message("Example usage of technical indicators.")
  
  # Load data
  collector <- CryptoDataCollector()
  btc_data <- collector$get_ohlcv_data("BTC-USD", period = "6mo")
  
  # Calculate indicators
  ti <- TechnicalIndicators(btc_data)
  data_with_indicators <- ti$add_all_indicators()
  
  # Get feature matrix
  features <- ti$get_feature_matrix()
  
  # Get summary
  summary <- ti$get_indicator_summary()
  print(summary)
  
  return(list(data = data_with_indicators, features = features, summary = summary))
} 